sklearn/decomposition/nmf.py

```diff
@@ -600,9 +741,264 @@ def _fit_coordinate_descent(X, W, H, tol=1e-4, max_iter=200, alpha=0.001,
+    """Compute Non-negative Matrix Factorization with Multiplicative Update
+
+    The objective function is beta_divergence(X, WH) and is minimized with an
+    alternating minimization of W and H. Each minimization is done with a
+    Multiplicative Update.
+
+    Parameters
+    ----------
+    X : array-like, shape (n_samples, n_features)
+        Constant input matrix.
+
+    W : array-like, shape (n_samples, n_components)
+        Initial guess for the solution.
+
+    H : array-like, shape (n_components, n_features)
+        Initial guess for the solution.
+
+    beta_loss : float or string, default 'frobenius'
+        String must be in {'frobenius', 'kullback-leibler', 'itakura-saito'}.
+        Beta divergence to be minimized, measuring the distance between X
+        and the dot product WH. Note that values different from 'frobenius'
+        (or 2) and 'kullback-leibler' (or 1) lead to significantly slower
+        fits.
+
+    n_iter : integer, default: 200
+        Number of iterations.
+
+    l1_reg_W : double, default: 0.
+        L1 regularization parameter for W.
+
+    l1_reg_H : double, default: 0.
+        L1 regularization parameter for H.
+
+    l2_reg_W : double, default: 0.
+        L2 regularization parameter for W.
+
+    l2_reg_H : double, default: 0.
+        L2 regularization parameter for H.
+
+    update_H : boolean, default: True
+        Set to True, both W and H will be estimated from initial guesses.
+        Set to False, only W will be estimated.
+
+    verbose : integer, default: 0
+        The verbosity level.
+
+    Returns
+    -------
+    W : array, shape (n_samples, n_components)
+        Solution to the non-negative least squares problem.
+
+    H : array, shape (n_components, n_features)
+        Solution to the non-negative least squares problem.
+
+    References
+    ----------
+    Fevotte, C., & Idier, J. (2011). Algorithms for nonnegative matrix
+    factorization with the beta-divergence. Neural Computation, 23(9).
+    """
+    beta_loss = _beta_loss_to_float(beta_loss)
+
+    # gamma for Maximization-Minimization (MM) algorithm [Fevotte 2011]
+    if beta_loss < 1:
+        gamma = 1. / (2. - beta_loss)
+    elif beta_loss > 2:
+        gamma = 1. / (beta_loss - 1.)
+    else:
+        gamma = 1.
+
+    start_time = time.time()
+    H_sum, HHt, XHt = None, None, None
+    for epoch in range(n_iter):
+        # update W
+        # H_sum, HHt and XHt are saved and reused if not update_H
+        delta_W, H_sum, HHt, XHt = _multiplicative_update_w(
+            X, W, H, beta_loss, l1_reg_W, l2_reg_W, gamma,
+            H_sum, HHt, XHt, update_H)
+        W *= delta_W
+
+        # update H
+        if update_H:
+            delta_H = _multiplicative_update_h(X, W, H, beta_loss, l1_reg_H,
+                                               l2_reg_H, gamma)
+            H *= delta_H
+
+            # These values will be recomputed since H changed
+            H_sum, HHt, XHt = None, None, None
```

Comment:
In #1348, the convergence check is `prev_error - error < tol`.
In #2540, the convergence check is `abs(1. - max_update) < tol`.

About computing the error, the cost is a bit high, yet I need to test it again.

About checking the maximum update, the update indeed converges to 1, but the criterion is rather poor if I recall correctly. After discussing with audio researchers in my group that use NMF a lot, a fixed number of iteration seems pretty classic.
Yet I can easily add this criterion if we consider it a better solution.